### PR TITLE
fix: Set default User-Agent header in Case.call

### PR DESCRIPTION
**Pull request by hebertjulio**
🚨Please review the [guidelines for contributing](https://github.com/schemathesis/schemathesis/blob/master/CONTRIBUTING.rst) to this repository.

### Description
Set default User-Agent header in Case.call.

### Checklist
- [ ] Created tests which fail without the change (if possible)
- [x] All tests passing
- [x] Added a changelog entry
- [ ] Extended the README / documentation, if necessary

### CODE REVIEW

File: src/schemathesis/models.py [hunk 1]
```diff
@@ -78,7 +79,7 @@ def formatted_path(self) -> str:
 
     def get_full_base_url(self) -> Optional[str]:
         """Create a full base url, adding "localhost" for WSGI apps."""
-        parts = urlsplit(self.base_url)
+        parts = urlsplit(self.base_url or "")
```
Comment (Stranger6667) [inline src/schemathesis/models.py:82]: It doesn't seem relevant to the topic of this PR. Why is it here?

File: src/schemathesis/models.py [hunk 2]
```diff
@@ -176,10 +177,12 @@ def call(
             close_session = True
         else:
             close_session = False
-
         data = self.as_requests_kwargs(base_url)
+        data["headers"] = data["headers"] or {}
         if headers is not None:
-            data["headers"] = {**(data["headers"] or {}), **headers}
+            data["headers"].update(headers)
```
Comment (Stranger6667) [inline src/schemathesis/models.py:183]: There is a problem with updating this key - it references the original `headers` dictionary (if it is not None) that is in the `self.headers` attribute. So if you are updating it, then it mutates that dictionary - if it is used in some other places in the test after `case.call`, then it might not be desired to have it mutated. I suggest adding `data["headers"] = data["headers"].copy()` few lines above (after this line - `data["headers"] = data["headers"] or {}`)

In the future, I plan to refactor it and make `as_requests_kwargs` and `as_werkzeug_kwargs` return copies to avoid unintentional mutations

File: src/schemathesis/models.py [hunk 3]
```diff
@@ -224,6 +227,11 @@ def call_wsgi(self, app: Any = None, headers: Optional[Dict[str, str]] = None, *
                 "Please, set `app` argument in the schema constructor or pass it to `call_wsgi`"
             )
         data = self.as_werkzeug_kwargs(headers)
+        data["headers"] = data["headers"] or {}
+        if headers is not None:
+            data["headers"].update(headers)
```
Comment (Stranger6667) [inline src/schemathesis/models.py:232]: These lines are not needed, since headers merging is handled in the `as_werkzeug_kwargs` method.

File: src/schemathesis/specs/graphql/schemas.py [hunk 4]
```diff
@@ -59,7 +59,7 @@ def base_path(self) -> str:
         return self._get_base_path()
 
     def _get_base_path(self) -> str:
-        return cast(str, urlsplit(self.location).path)
+        return cast(str, urlsplit(self.location or "").path)
```
Comment (Stranger6667) [inline src/schemathesis/specs/graphql/schemas.py:62]: This change doesn't seem relevant to the topic of this PR. Why is it changed here?

File: test/cli/test_cassettes.py [hunk 5]
```diff
@@ -153,14 +154,16 @@ async def test_replay(openapi_version, cli, schema_url, app, reset_app, cassette
...
+    headers = {"User-Agent": USER_AGENT}
+    response = requests.get(f"{base_url}/success", cookies={"foo": "bar", "baz": "spam"}, headers=headers)
     request = Request.from_prepared_request(response.request)
     serialized = {
         "uri": request.uri,
         "method": request.method,
         "headers": request.headers,
         "body": {"encoding": "utf-8", "base64_string": request.body},
     }
+    assert USER_AGENT in serialized["headers"]["User-Agent"]
```
Comment (Stranger6667) [inline test/cli/test_cassettes.py:166]:
```suggestion
    assert serialized["headers"]["User-Agent"] == USER_AGENT
```

I think that comparing the whole string is better than comparing if a substring is present in this case

File: test/test_models.py [hunk 6]
```diff
@@ -17,9 +18,10 @@ def test_path(swagger_20):
 @pytest.mark.parametrize("override", (False, True))
 @pytest.mark.parametrize("converter", (lambda x: x, lambda x: x + "/"))
 def test_as_requests_kwargs(override, server, base_url, swagger_20, converter):
+    headers = {"User-Agent": USER_AGENT}
     base_url = converter(base_url)
     endpoint = Endpoint("/success", "GET", {}, swagger_20)
-    kwargs = {"endpoint": endpoint, "cookies": {"TOKEN": "secret"}}
+    kwargs = {"endpoint": endpoint, "cookies": {"TOKEN": "secret"}, "headers": headers}
```
Comment (Stranger6667) [inline test/test_models.py:24]: I'd rather revert this change and verify that by default headers are updated with the "User-Agent" header

**Review by Stranger6667 — CHANGES_REQUESTED** [review]

File: test/cli/test_cassettes.py (hunk 5 above)
Comment (hebertjulio) [reply test/cli/test_cassettes.py:166]: It true, but serialized["headers"]["User-Agent"] is a tuple.

Comment (Stranger6667) [reply test/cli/test_cassettes.py:166]: You are right, my bad

File: src/schemathesis/models.py (hunk 1 above)
Comment (hebertjulio) [reply src/schemathesis/models.py:82]: My PRs are failing in mypy due to an optional value being passed to this function

File: src/schemathesis/specs/graphql/schemas.py (hunk 4 above)
Comment (hebertjulio) [reply src/schemathesis/specs/graphql/schemas.py:62]: My PRs are failing in mypy due to an optional value being passed to this function

File: src/schemathesis/models.py (hunk 1 above)
Comment (Stranger6667) [reply src/schemathesis/models.py:82]: Aha, I see. It might be caused by the recent changes in the pre-commit config - I will fix it soon

File: src/schemathesis/models.py (hunk 3 above)
Comment (hebertjulio) [reply src/schemathesis/models.py:232]: Yes, it  true, I will remove these lines.

File: src/schemathesis/models.py (hunk 1 above)
Comment (hebertjulio) [reply src/schemathesis/models.py:82]: I waiting you fix it?

File: src/schemathesis/models.py (hunk 2 above)
Comment (hebertjulio) [reply src/schemathesis/models.py:183]: Ok

File: src/schemathesis/models.py (hunk 1 above)
Comment (Stranger6667) [reply src/schemathesis/models.py:82]: Gimme 5 minutes and then you can rebase your branch

Comment (Stranger6667) [reply src/schemathesis/models.py:82]: Ok, now it should be fine - I reverted the change of the default Python version for pre-commit. Let me know if it works for you

File: test/cli/test_cassettes.py (hunk 5 above)
Comment (hebertjulio) [reply test/cli/test_cassettes.py:166]: No problem :)

File: test/test_models.py (hunk 6 above)
Comment (hebertjulio) [reply test/test_models.py:24]: as_requests_kwargs dont set default user agent
